docs/auth/oidc.md

```diff
@@ -0,0 +1,231 @@
+
+An example of such would be when you use an auth provider from a library installed with
+NPM, or any other library repository, you would import the API ref from the library.
+
+```ts
+export const azureOIDCAuthApiRef: ApiRef<
+  OpenIdConnectApi & ProfileInfoApi & BackstageIdentityApi & SessionApi
+> = createApiRef({
+  id: 'core.auth.azureOIDC',
+});
+```
+
+Please note a few things, the ID can be anything you want as long as it doesn't conflict
+with other refs, also we're exporting this reference, as well as the `typings`, we need to
+be able to import this reference anywhere in the app, and the `typings` will tell typescript
+what instance we're getting from DI when injecting the API. In this case we are defining
+an API for authentication, so we tell TS that this instance complies with 4 API
+interfaces:
+
+- The OICD API that will handle authentication.
+- Profile API for requesting user profile info from the auth provider in question.
+- Backstage identity API to handle and associate the user profile with backstage identity.
+- Session API, to handle the session the user will have while logged in.
+
+### The API Factory
+
+A factory is a function that can take some parameters or dependencies and return an
+instance of something, in our case it will be a function that requests some backstage
+APIs and use them to create an instance of an OIDC provider.
+
+Please note that this function only runs (creates the instance) when somewhere else in
+the app you request the DI to give you an instance of the OIDC provider using the API ref
+defined above, and the DI will only run this function the first time, from then on any
+other DI injection will just receive the same instance created the first time, basically
+the instance is cached by the DI library, a singleton.
+
+Let's add our OIDC factory to the APIs array in the `packages/app/src/apis.ts` file:
+
+```diff
++ import { OAuth2 } from '@backstage/core-app-api';
+
+export const apis: AnyApiFactory[] = [
++   createApiFactory({
++    api: azureOIDCAuthApiRef,
++    deps: {
++      discoveryApi: discoveryApiRef,
++      oauthRequestApi: oauthRequestApiRef,
++      configApi: configApiRef,
++    },
++    factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
++      OAuth2.create({
++        discoveryApi,
++        oauthRequestApi,
++        provider: {
++          id: 'oidc', // This has to be 'oidc' or OAuth2 will not use oidc protocol
++          title: 'OIDC provider',
++          icon: () => null,
++        },
++        environment: configApi.getOptionalString('auth.environment'),
++        defaultScopes: [
++          'openid',
++          'profile',
++          'email',
++        ],
++      }),
++  }),
+
+```
+
+Please note we're importing the `OAuth2` class from `@backstage/core-app-api` effectively
+delegating the authentication to it (yes it can handle OIDC as well). Also we're using
+the `oidc` ID to tell `OAuth2` to use the OIDC protocol, and added the default scopes to
```

Comment:
This isn't really true. All it's doing is telling it to use the `/api/auth/oidc/` endpoint

Reply:
Maybe it depends on the provider, for Azure I had to put `oidc` so the URL would be that one, haven't tried other providers yet, what do Google and AWS use for the redirect URI when using OIDC protocol?

Reply:
Not sure if it has to be `oidc` to match the OIDC provider instance that is defined in [oidc/provider.ts#L206](https://github.com/backstage/backstage/blob/master/plugins/auth-backend/src/providers/oidc/provider.ts#L206) and is made available in the application in [providers/providers.ts#L75](https://github.com/backstage/backstage/blob/master/plugins/auth-backend/src/providers/providers.ts#L75).

Maybe I can just use `azureOIDC` as the ID in there as well, since we're defining a new provider factory to use a custom sign in resolver, I'll try to run this locally and see if it works.

Reply:
Ok just tested stuff and the only parts that have to match are the `providerFactories` key for the resolver, the provider's `id` inside the API factory, the yaml provider key, and the callback URI provider segment, will change to `my-provider` and mention this requirement.